org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java

```diff
@@ -1933,4 +1968,108 @@ public void setAndroidSupportEnabled(boolean androidSupportEnabled) {
 		this.androidSupportEnabled = androidSupportEnabled;
 	}
 
+	public List<String> getNonnullTypes() {
+		return this.nonnullTypes;
+	}
+
+	public void setNonnullTypes(List<String> nonnullTypes) {
+		this.nonnullTypes = nonnullTypes;
+	}
+
+	public List<String> getNullableTypes() {
+		return this.nullableTypes;
+	}
+
+	public void setNullableTypes(List<String> nullableTypes) {
+		this.nullableTypes = nullableTypes;
+	}
+
+	public boolean isAnnotationNullAnalysisEnabled() {
+		return !this.nonnullTypes.isEmpty() || !this.nullableTypes.isEmpty();
+	}
+
+	/**
+	 * @return whether the options are changed or not
+	 */
+	public boolean updateAnnotationNullAnalysisOptions() {
+		boolean isChanged = false;
+		for (IJavaProject javaProject : ProjectUtils.getJavaProjects()) {
+			isChanged |= updateAnnotationNullAnalysisOptions(javaProject);
+		}
+		return isChanged;
+	}
+
+	/**
+	 * @param javaProject the java project to update the annotation-based null analysis options
+	 * @return whether the options of the given project are changed or not
+	 */
+	public boolean updateAnnotationNullAnalysisOptions(IJavaProject javaProject) {
+		Map<String, String> projectOptions = javaProject.getOptions(true);
+		if (projectOptions == null) {
+			return false;
+		}
+		String nonnullType = getAnnotationType(javaProject, this.nonnullTypes, builtinNonnullTypes);
+		String nullableType = getAnnotationType(javaProject, this.nullableTypes, builtinNullableTypes);
+		if (nullableType == null && nonnullType == null) {
+			disableAnnotationNullAnalysis(projectOptions);
+		} else {
+			enableAnnotationNullAnalysis(projectOptions, nonnullType, nullableType);
+		}
+		boolean isChanged = !(javaProject.getOptions(true).equals(projectOptions));
+		javaProject.setOptions(projectOptions);
+		return isChanged;
+	}
+
+	private String getAnnotationType(IJavaProject javaProject, List<String> annotationTypes, Map<String, String> builtinAnnotationTypes) {
+		try {
+			ClasspathResult result = ProjectCommand.getClasspathsFromJavaProject(javaProject, new ProjectCommand.ClasspathOptions());
+			for (String annotationType : annotationTypes) {
+				if (builtinAnnotationTypes.keySet().contains(annotationType)) {
+					// for known types, check the classpath to achieve a better performance
+					if (javaProject.getElementName().equals(ProjectsManager.DEFAULT_PROJECT_NAME)) {
+						continue;
+					}
+					for (String classpath : result.classpaths) {
+						if (classpath.contains(builtinAnnotationTypes.get(annotationType))) {
```

Comment:
This check fails for me where it should have succeeded.

My `classpath` had `/home/rgrunber/.m2/repository/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar`

The expected match should be against `com.google.code.findbugs/jsr305` but clearly they won't match. You may need to adjust this if you want to check the presence for the artifact. It also becomes dependent on the format of the locally downloaded artifact but I guess we can make this assumption for now.

Also what if someone has an invisible project where `lib/foo.jar` is actually the `jsr305` artifact ? I think the only way to really know the annotation type exists, is through `findType`, but it can be expensive.

Reply:
Another way to check this is to just check the artifact id (jsr305), but it might run into some issues since the artifact id of type `org.jetbrains.annotations.NotNull` is `annotations`, I think it's quite common and it might lead some mistakes. Do you think it makes sense? 

Or we can make the classpath jar list dynamic, when `findType` find the correct type we can add it to the list so that we need only to run it once. But the problem is how to get the classpath jar of IType (maybe from IClassfile), and if we need to do it.

Reply:
Given any java element, you should be able to do `element.getAncestor(IJavaElement.PACKAGE_FRAGMENT_ROOT).getPath()` .

Reply:
it looks like we can do this, via `javaProject.findType(${typeName}).getClassFile().getPath()`, and store it into the cache.

Reply:
@CsCherrYY Your current approach of getting the path has problems.

A quick check for spring-core on Windows, it's `.m2\repository\org\springframework\spring-core\5.3.21\spring-core-5.3.21.jar`, not `org.springframework\spring-core\...`.

Reply:
> Given any java element (IJavaElement), you should be able to do element.getAncestor(IJavaElement.PACKAGE_FRAGMENT_ROOT).getPath() .

Just tested and it works 👍

So here is a complete new proposal:

1. As the default list, it should contain both of two styles (e.g., `org.springframework\spring-core\...` and `org\springframework\spring-core\...`)
2. Adopt dynamic list, once a type founded by `findType()`, add it and corresponding classpath jar into the list, to avoid findType for it again.

@rgrunber @testforstephen WDYT?

Reply:
I'm curious what kind of case is `org.springframework\spring-core\...`?

Reply:
Looks like maven is saving the groupid into separate folders, but gradle is just saving to one directory directly.

Reply:
> > Given any java element (IJavaElement), you should be able to do element.getAncestor(IJavaElement.PACKAGE_FRAGMENT_ROOT).getPath() .
> 
> Just tested and it works 👍
> 
> So here is a complete new proposal:
> 
> 1. As the default list, it should contain both of two styles (e.g., `org.springframework\spring-core\...` and `org\springframework\spring-core\...`)
> 2. Adopt dynamic list, once a type founded by `findType()`, add it and corresponding classpath jar into the list, to avoid findType for it again.
> 
> @rgrunber @testforstephen WDYT?

LGTM